src/crimson/os/seastore/omap_manager/btree/omap_btree_node_impl.h

```diff
@@ -28,20 +30,88 @@ namespace crimson::os::seastore::omap_manager {
 
 struct OMapInnerNode
   : OMapNode,
-    StringKVInnerNodeLayout {
+    StringKVInnerNodeLayout,
+    ParentNode<OMapInnerNode, std::string>,
+    ChildNode<OMapInnerNode, OMapInnerNode, std::string> {
   using OMapInnerNodeRef = TCachedExtentRef<OMapInnerNode>;
-  using internal_iterator_t = const_iterator;
+  using internal_const_iterator_t = const_iterator;
+  using internal_iterator_t = iterator;
+  using parent_node_t = ParentNode<OMapInnerNode, std::string>;
+  using base_child_t = BaseChildNode<OMapInnerNode, std::string>;
+  using child_node_t = ChildNode<OMapInnerNode, OMapInnerNode, std::string>;
+  static constexpr uint32_t CHILD_VEC_UNIT = 128;
 
   explicit OMapInnerNode(ceph::bufferptr &&ptr)
-    : OMapNode(std::move(ptr)) {
-    this->set_layout_buf(this->get_bptr().c_str());
+    : OMapNode(std::move(ptr)),
+      StringKVInnerNodeLayout(get_bptr().c_str()),
+      parent_node_t(0) {
+    this->parent_node_t::sync_children_capacity();
   }
   // Must be identical with OMapInnerNode(ptr) after on_fully_loaded()
   explicit OMapInnerNode(extent_len_t length)
-    : OMapNode(length) {}
+    : OMapNode(length),
+      StringKVInnerNodeLayout(nullptr),
+      parent_node_t(0) {}
   OMapInnerNode(const OMapInnerNode &rhs)
-    : OMapNode(rhs) {
-    this->set_layout_buf(this->get_bptr().c_str());
+    : OMapNode(rhs),
+      StringKVInnerNodeLayout(get_bptr().c_str()),
+      parent_node_t(0) {
+    this->parent_node_t::sync_children_capacity();
+  }
+
+  iterator begin() {
+    return iter_begin();
+  }
+
+  iterator end() {
+    return iter_end();
+  }
+
+  void do_on_rewrite(Transaction &t, LogicalCachedExtent &extent) final {
+    auto &ext = static_cast<OMapInnerNode&>(extent);
+    this->parent_node_t::on_rewrite(t, ext);
+    auto &other = static_cast<OMapInnerNode&>(extent);
+    this->init_meta(other.get_begin(), other.get_end());
+    this->sync_children_capacity();
+  }
+
+  void prepare_commit() final {
+    this->parent_node_t::prepare_commit();
+    if (is_rewrite() && !is_btree_root()) {
+      auto &prior = *get_prior_instance()->template cast<OMapInnerNode>();
+      if (prior.base_child_t::has_parent_tracker()) {
+	// unlike fixed-kv nodes, rewriting child nodes of the omap tree
+	// won't affect parent nodes, so we have to manually take prior
+	// instances' parent trackers here.
+	this->child_node_t::take_parent_from_prior();
+      }
+    }
+  }
+
+  void do_on_replace_prior() final {
+    this->parent_node_t::on_replace_prior();
+    if (!this->is_btree_root()) {
+      // XXX: On startup, parent/child nodes may not be linked,
+      //      we need to tolerate this case.
```

Comment:
IIUC in order to mutate an omap extent, it needs to be looked up from the root, IIUC it should be linked during the lookup.

Probably I'm missing something, upon committing a mutation pending omap extent, why the prior omap extent may not be linked yet?

Reply:
If I'm correct above, I assume the additional `has_parent_tracker()` is only necessary in the destructors.

And we can also move the related comments there.

Reply:
You're right, I've removed this comment and turned to code into `assert(prior.has_parent_tracker()`